### PR TITLE
Fix/refactor fragment edit

### LINKED ISSUE #41: Refactor EditFragment
- [x] refactor functions to activate/deactivate the LocationPicker ('activateLocationPickerElements' / 'deactivateLocationPickerElements'
- [x] modularize code in EditFragment, so that the file becomes smaller
- [x] consider to encapsulate the verification the contact email attribute into a seperate function (not null, not blank, not empty)
- [x] consider to encapsulate the verification of both isShared and location into a seperate function (location shall not be null, if isShared is set to true)
- [x] Add ability to reset the best by date
- [x] make Unit dropdown arrow clickable
- [x] remove usages of synthetic package

### PR DESCRIPTION
**Pull request by henrijaeger**
Closes  #41, #46.

- [x] refactor functions to activate/deactivate the LocationPicker ('activateLocationPickerElements' / 'deactivateLocationPickerElements'
- [x] modularize code in EditFragment, so that the file becomes smaller
- [x] consider to encapsulate the verification the contact email attribute into a seperate function (not null, not blank, not empty)
- [x] consider to encapsulate the verification of both isShared and location into a seperate function (location shall not be null, if isShared is set to true)
- [x] Add ability to reset the best by date (+ other fields with just a tap)
- [x] make Unit dropdown arrow clickable
- [x] remove usages of synthetic package

### CODE REVIEW

**Review by AlexTerhar — APPROVED** [review]

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 1]
```diff
@@ -100,437 +100,264 @@ class EditFragment : Fragment() {
...
     }
 
-    private fun saveNewItem() {
-        // TODO: put the following condition into a separate function
-        if (itemIsSharedSwitch.isChecked && location == null) {
-            displayAlertOnSaveSharedItemWithoutLocation()
-            itemAddressTextInputLayout.editText?.setText("")
+    private fun setUpSaveButton() {
+        if (!ADD_ITEM_MODE) {
+            binding.itemSaveButton.isVisible = false
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/EditFragment.kt:277]: Return here and remove else

File: app/src/main/java/app/wefridge/wefridge/EditFragment.kt [hunk 2]
```diff
@@ -100,437 +100,264 @@ class EditFragment : Fragment() {
...
         } else {
+            binding.itemSaveButton.setOnClickListener {
+                binding.itemAddressTextInputLayout.clearFocus()
+
+                try {
+                    saveItem(
+                        callbackOnSuccess = {
+                            Toast.makeText(
+                                requireContext(),
+                                "Item was successfully saved!",
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/EditFragment.kt:287]: Strings!

File: app/src/main/java/app/wefridge/wefridge/model/Unit.kt [hunk 3]
```diff
@@ -20,6 +21,17 @@ enum class Unit(val value: Int, @StringRes val _display: Int) {
...
         fun getByValue(value: Int?) = VALUES.firstOrNull { it.value == value }
 
+        fun getByString(symbol: String, callingFragment: Fragment): Unit? {
+            return when (symbol) {
+                callingFragment.getString(R.string.item_unit_gram) -> GRAM
+                callingFragment.getString(R.string.item_unit_kilogram) -> KILOGRAM
+                callingFragment.getString(R.string.item_unit_liter) -> LITER
+                callingFragment.getString(R.string.item_unit_milliliter) -> MILLILITER
+                callingFragment.getString(R.string.item_unit_ounce) -> OUNCE
+                callingFragment.getString(R.string.item_unit_piece) -> PIECE
```
Comment (frcroth) [inline app/src/main/java/app/wefridge/wefridge/model/Unit.kt:31]: Possibly this can be simplified by finding an enum whose display value matches? Probably @PassiDel has a good idea here, so that only the enum contains mapping information, not the methods

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -103,6 +104,35 @@
...
+    <string name="ad_msg_contact_email_missing_on_destroy">In order to share an Item, you have to provide a contact email other users can use to reach out to you. Please state your email address in the settings tab.</string>
+    <string name="ad_msg_error_parsing_address_string">There was an error while we tried to match the location you\'ve entered to real coordinates. Please correct the address you\'ve entered or disable the sharing option. You may also want to check your internet connection and try again.</string>
+    <string name="ad_msg_error_parsing_address_string_on_destroy">There was an error while we tried to match the location you\'ve entered to real coordinates. Therefore, we disabled the sharing option. Please correct the address you\'ve entered in order to share the foodstuff. You may also want to check your internet connection and try again.</string>
+    <string name="ad_msg_no_network">There was an error while we tried to match the location coordinates to a real address. Therefore, we disabled the sharing option. Please check your internet connection and try again.</string>
```
Comment (frcroth) [inline app/src/main/res/values/strings.xml:125]: Why this again? Isn't this a duplicate of 50 and following?

**Review by frcroth — COMMENTED** [review]
Overall looking very good and sensible, big improvement. Some notes here that may be improved. I am confused about strings.xml, otherwise everything is minor and not merge blocking.

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (henrijaeger) [reply app/src/main/res/values/strings.xml:125]: Oh, that must have been my mistake while resolving merge conflicts. I will fix that!

File: app/src/main/res/layout/fragment_edit.xml [hunk 5]
```diff
@@ -35,19 +38,22 @@
...
-                <com.google.android.material.textfield.TextInputEditText
+                <app.wefridge.wefridge.EditItemEditText
                     android:layout_width="match_parent"
                     android:layout_height="wrap_content"
                     android:layout_marginEnd="8dp"
                     android:inputType="number"/>
-            </com.google.android.material.textfield.TextInputLayout>
+            </app.wefridge.wefridge.EditItemTextInputLayout>
 
             <com.google.android.material.textfield.TextInputLayout
```
Comment (PassiDel) [inline app/src/main/res/layout/fragment_edit.xml:58]: The unit dropdown has a wrong height

**Review by PassiDel — CHANGES_REQUESTED** [review]
I hope I got everything.

TL;DR:
- Use Context instead of Fragment, for AlertDialogs
- Store Unit ID in MenuItem, Single source of truth (Enum Unit)

File: app/src/main/java/app/wefridge/wefridge/model/LocationController.kt [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+import com.google.android.gms.location.FusedLocationProviderClient
+import com.google.android.gms.location.LocationRequest
+import com.google.android.gms.location.LocationServices
+import com.google.android.gms.tasks.CancellationTokenSource
+import com.google.firebase.firestore.GeoPoint
+import java.io.IOException
+import java.util.*
+import kotlin.Unit
+
+class LocationController(private val servedFragment: Fragment, private val callbackOnPermissionDenied: () -> Unit, private val callbackForPermissionRationale: () -> Unit, private val callbackOnDeterminationFailed: () -> Unit, private val callbackOnSuccess: (geoPoint: GeoPoint) -> Unit) {
```
[the remark this answers is not in the document]
Comment (henrijaeger) [reply app/src/main/java/app/wefridge/wefridge/model/LocationController.kt:21]: I tried that, but it causes more redundancy in the EditFragment. Some callbacks must then be defined twice (for both getCurrentLocation and another function which replaces the Initializer). Working with objects, the LocationController's behaviour can be defined once at initialization and reused after that in various functions.

**Comment by henrijaeger** [comment]
@frcroth @PassiDel Thank you very much for your thoughtful feedback!
